src/Features/Core/Portable/ForeachToFor/AbstractForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,346 @@
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(context.Span.Start);
+
+            var foreachStatement = GetForeachStatement(token);
+            if (foreachStatement == null)
+            {
+                return;
+            }
+
+            var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var semanticFact = document.GetLanguageService<ISemanticFactsService>();
+            var foreachInfo = GetForeachInfo(semanticFact, model, foreachStatement, cancellationToken);
+            if (foreachInfo == null)
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(
+                new ForeachToForCodeAction(
+                    FeaturesResources.Convert_foreach_to_for,
+                    c => ConvertForeachToForAsync(document, foreachInfo, c)));
+        }
+
+        protected string CreateUniqueMethodName(SemanticModel model, SyntaxNode contextNode, string baseName)
+        {
+            Contract.ThrowIfNull(contextNode);
+            Contract.ThrowIfNull(baseName);
+
+            return NameGenerator.GenerateUniqueName(baseName, string.Empty,
+               name => model.LookupSymbols(contextNode.SpanStart, /*container*/null, name).Length == 0);
+        }
+
+        protected T AddElasticAnnotation<T>(T node, SyntaxTrivia elasticTrivia) where T : SyntaxNode
+        {
+            var first = node.GetFirstToken();
+            node = node.ReplaceToken(first, first.WithPrependedLeadingTrivia(elasticTrivia));
+
+            var last = node.GetLastToken();
+            return node.ReplaceToken(last, last.WithAppendedTrailingTrivia(elasticTrivia));
+        }
+
+        protected T AddRenameAnnotation<T>(T node, string name) where T : SyntaxNode
+        {
+            var token = node.DescendantTokens().FirstOrDefault(t => t.Text == name);
+            if (token == default(SyntaxToken))
+            {
+                return node;
+            }
+
+            return node.ReplaceToken(token, token.WithAdditionalAnnotations(RenameAnnotation.Create()));
+        }
+
+        private ForeachInfo GetForeachInfo(ISemanticFactsService semanticFact, SemanticModel model, SyntaxNode foreachStatement, CancellationToken cancellationToken)
+        {
+            var operation = model.GetOperation(foreachStatement, cancellationToken) as IForEachLoopOperation;
+            if (operation == null || operation.Locals.Length != 1)
+            {
+                return null;
+            }
+
+            var foreachVariable = operation.Locals[0];
+            if (foreachVariable == null)
```

Comment:
how does this happen?

Reply:
when there is no local in foreach( in )

Reply:
Interesting.  I'm not sure if i saw any tests for that.  If they exist, great.  If not, please add.